File: python/tests/test_mock.py
"""Tests for MockRemit — all Wallet methods tested against the in-memory mock."""

import pytest
import pytest_asyncio

from remitmd.errors import (
    BountyNotFound,
    InsufficientBalance,
    InvalidState,
    TabLimitExceeded,
    TabNotFound,
)
from remitmd.models.common import (
    BountyStatus,
    DepositStatus,
    DisputeStatus,
    EscrowStatus,
    StreamStatus,
    TabStatus,
)
from remitmd.models.invoice import Invoice
from remitmd.testing.mock import MockRemit


@pytest.fixture
def mock() -> MockRemit:
    return MockRemit()


@pytest.fixture
def payer(mock: MockRemit):
    return mock.create_wallet(balance=1000.0)


@pytest.fixture
def payee(mock: MockRemit):
    return mock.create_wallet(balance=0.0)


# ─── Direct payment ───────────────────────────────────────────────────────────

@pytest.mark.asyncio
async def test_pay_direct_transfers_funds(mock, payer, payee):
    await payer.pay_direct(payee.address, 50.0, memo="hello")

    assert await payer.balance() == 950.0
    assert await payee.balance() == 50.0


@pytest.mark.asyncio
async def test_pay_direct_insufficient_balance(mock, payer, payee):
    with pytest.raises(InsufficientBalance):
        await payer.pay_direct(payee.address, 2000.0)


# ─── Escrow ───────────────────────────────────────────────────────────────────

@pytest.mark.asyncio
async def test_escrow_full_lifecycle(mock, payer, payee):
    invoice = Invoice(to=payee.address, amount=100.0, memo="task")
    tx = await payer.pay(invoice)
    assert tx.invoice_id is not None
    assert await payer.balance() == 900.0

    await payer.release_escrow(tx.invoice_id)
    assert await payee.balance() == 100.0


@pytest.mark.asyncio
async def test_escrow_cancel_refunds_payer(mock, payer, payee):
    invoice = Invoice(to=payee.address, amount=100.0)
    tx = await payer.pay(invoice)
    await payer.cancel_escrow(tx.invoice_id)
    assert await payer.balance() == 1000.0
    assert await payee.balance() == 0.0


@pytest.mark.asyncio
async def test_escrow_double_release_raises(mock, payer, payee):
    invoice = Invoice(to=payee.address, amount=100.0)
    tx = await payer.pay(invoice)
    await payer.release_escrow(tx.invoice_id)

    with pytest.raises(InvalidState):
        await payer.release_escrow(tx.invoice_id)


@pytest.mark.asyncio
async def test_escrow_claim_start_changes_status(mock, payer, payee):
    invoice = Invoice(to=payee.address, amount=50.0)
    tx = await payer.pay(invoice)
    await payee.claim_start(tx.invoice_id)

    escrow = mock._state.escrows[list(mock._state.escrows.keys())[-1]]
    assert escrow.status == EscrowStatus.active


# ─── Tabs ─────────────────────────────────────────────────────────────────────

@pytest.mark.asyncio
async def test_tab_open_and_close(mock, payer, payee):
    tab = await payer.open_tab(payee.address, limit=10.0, per_unit=0.01)
    assert tab.status == TabStatus.open
    assert await payer.balance() == 990.0  # limit reserved

    await payer.charge_tab(tab.id, units=100)  # 100 * $0.01 = $1.00

    await payer.close_tab(tab.id)
    # payee gets $1.00, payer gets back $9.00
    assert await payee.balance() == 1.0
    assert await payer.balance() == 999.0


@pytest.mark.asyncio
async def test_tab_charge_exceeds_limit(mock, payer, payee):
    tab = await payer.open_tab(payee.address, limit=1.0, per_unit=0.01)
    with pytest.raises(TabLimitExceeded):
        await payer.charge_tab(tab.id, units=200)  # 200 * $0.01 = $2.00 > $1.00 limit


@pytest.mark.asyncio
async def test_tab_close_unknown_id(mock, payer):
    with pytest.raises(TabNotFound):
        await payer.close_tab("nonexistent_id")


@pytest.mark.asyncio
async def test_tab_double_close_raises(mock, payer, payee):
    tab = await payer.open_tab(payee.address, limit=5.0, per_unit=0.10)
    await payer.close_tab(tab.id)
    with pytest.raises(InvalidState):
        await payer.close_tab(tab.id)


# ─── Streams ──────────────────────────────────────────────────────────────────

@pytest.mark.asyncio
async def test_stream_open_and_close(mock, payer, payee):
    stream = await payer.open_stream(payee.address, rate=1.0, max_duration=60)
    assert stream.status == StreamStatus.active

    mock.advance_time(30)  # 30 seconds pass
    await payer.close_stream(stream.id)

    # Payee should get roughly 30 * $1.00 = $30 (mock uses time.time() so may vary)
    payee_balance = await payee.balance()
    assert payee_balance >= 0.0
    # Payer got remainder back
    assert await payer.balance() + payee_balance == pytest.approx(1000.0, abs=1.0)


# ─── Bounties ─────────────────────────────────────────────────────────────────

@pytest.mark.asyncio
async def test_bounty_post_and_award(mock, payer, payee):
    bounty = await payer.post_bounty(
        amount=50.0, task="write tests", deadline=mock.now() + 3600
    )
    assert bounty.status == BountyStatus.open
    assert await payer.balance() == 950.0

    await payee.submit_bounty(bounty.id, evidence_uri="ipfs://Qm...")
    await payer.award_bounty(bounty.id, winner=payee.address)

    assert await payee.balance() == 50.0
    assert mock._state.bounties[bounty.id].status == BountyStatus.awarded


@pytest.mark.asyncio
async def test_bounty_award_unknown(mock, payer, payee):
    with pytest.raises(BountyNotFound):
        await payer.award_bounty("nonexistent", winner=payee.address)


@pytest.mark.asyncio
async def test_bounty_double_award_raises(mock, payer, payee):
    bounty = await payer.post_bounty(25.0, "task", deadline=mock.now() + 3600)
    await payer.award_bounty(bounty.id, winner=payee.address)
    with pytest.raises(InvalidState):
        await payer.award_bounty(bounty.id, winner=payee.address)


# ─── Deposits ─────────────────────────────────────────────────────────────────

@pytest.mark.asyncio
async def test_deposit_locked_and_returned(mock, payer, payee):
    deposit = await payer.place_deposit(payee.address, amount=20.0, expires=3600)
    assert deposit.status == DepositStatus.locked
    assert await payer.balance() == 980.0

    await payee.return_deposit(deposit.id)
    assert await payer.balance() == 1000.0
    assert mock._state.deposits[deposit.id].status == DepositStatus.returned


@pytest.mark.asyncio
async def test_deposit_forfeited(mock, payer, payee):
    deposit = await payer.place_deposit(payee.address, amount=20.0, expires=3600)
    await payee.forfeit_deposit(deposit.id)
    assert await payee.balance() == 20.0
    assert mock._state.deposits[deposit.id].status == DepositStatus.forfeited


# ─── Disputes ─────────────────────────────────────────────────────────────────

@pytest.mark.asyncio
async def test_file_dispute(mock, payer, payee):
    invoice = Invoice(to=payee.address, amount=100.0)
    tx = await payer.pay(invoice)
    dispute = await payer.file_dispute(
        invoice_id=tx.invoice_id,
        reason="non_delivery",
        details="Work was not completed",
        evidence_uri="ipfs://Qm...",
    )
    assert dispute.status == DisputeStatus.open
    assert dispute.invoice_id == tx.invoice_id


# ─── Forced errors ────────────────────────────────────────────────────────────

@pytest.mark.asyncio
async def test_set_behavior_forces_error(mock, payer, payee):
    from remitmd.errors import RemitError
    mock.set_behavior(payer.address, "RATE_LIMIT_EXCEEDED")
    with pytest.raises(RemitError) as exc_info:
        await payer.pay_direct(payee.address, 1.0)
    assert exc_info.value.code == "RATE_LIMIT_EXCEEDED"


@pytest.mark.asyncio
async def test_forced_error_is_consumed_once(mock, payer, payee):
    mock.set_behavior(payer.address, "SERVER_ERROR")
    from remitmd.errors import RemitError
    with pytest.raises(RemitError):
        await payer.pay_direct(payee.address, 1.0)
    # Second call should succeed
    tx = await payer.pay_direct(payee.address, 1.0)
    assert tx.tx_hash is not None


# ─── Testnet faucet ───────────────────────────────────────────────────────────

@pytest.mark.asyncio
async def test_request_testnet_funds(mock, payer):
    before = await payer.balance()
    await payer.request_testnet_funds()
    after = await payer.balance()
    assert after == before + 100.0


# ─── Webhooks ─────────────────────────────────────────────────────────────────

@pytest.mark.asyncio
async def test_register_webhook(mock, payer):
    wh = await payer.register_webhook(
        url="https://example.com/hook",
        events=["escrow.funded", "escrow.released"],
    )
    assert wh.active is True
    assert "escrow.funded" in wh.events

File: typescript/tests/test_mock.ts
import { describe, it, beforeEach } from "node:test";
import assert from "node:assert/strict";

import { MockRemit } from "../src/testing/mock.js";
import { InsufficientBalanceError } from "../src/errors.js";

describe("MockRemit", () => {
  let mock: MockRemit;

  beforeEach(() => {
    mock = new MockRemit();
  });

  it("createWallet returns wallet with correct balance", async () => {
    const wallet = mock.createWallet(500);
    const bal = await wallet.balance();
    assert.equal(bal, 500);
  });

  it("payDirect transfers funds", async () => {
    const payer = mock.createWallet(100);
    const payee = mock.createWallet(0);
    const tx = await payer.payDirect(payee.address, 30);
    assert.equal(tx.status, "confirmed");
    assert.equal(await payer.balance(), 70);
    assert.equal(await payee.balance(), 30);
  });

  it("payDirect throws on insufficient balance", async () => {
    const payer = mock.createWallet(10);
    const payee = mock.createWallet(0);
    await assert.rejects(
      () => payer.payDirect(payee.address, 100),
      InsufficientBalanceError,
    );
  });

  it("escrow: fund, release", async () => {
    const payer = mock.createWallet(200);
    const payee = mock.createWallet(0);

    const tx = await payer.pay({
      id: "inv-001",
      from: payer.address,
      to: payee.address,
      amount: 100,
      chain: "base",
      status: "pending",
      paymentType: "escrow",
      createdAt: Math.floor(Date.now() / 1000),
    });

    assert(tx.txHash);
    assert.equal(await payer.balance(), 100); // deducted on fund

    const releaseTx = await payer.releaseEscrow("inv-001");
    assert.equal(releaseTx.status, "confirmed");
    assert.equal(await payee.balance(), 100);
  });

  it("escrow: fund, cancel - refunds payer", async () => {
    const payer = mock.createWallet(200);
    const payee = mock.createWallet(0);

    await payer.pay({
      id: "inv-002",
      from: payer.address,
      to: payee.address,
      amount: 50,
      chain: "base",
      status: "pending",
      paymentType: "escrow",
      createdAt: Math.floor(Date.now() / 1000),
    });

    assert.equal(await payer.balance(), 150);
    await payer.cancelEscrow("inv-002");
    assert.equal(await payer.balance(), 200); // refunded
    assert.equal(await payee.balance(), 0);
  });

  it("tab: open, close - refunds unspent", async () => {
    const payer = mock.createWallet(100);
    const payee = mock.createWallet(0);

    const tab = await payer.openTab({ to: payee.address, limit: 50, perUnit: 5 });
    assert.equal(tab.status, "open");
    assert.equal(await payer.balance(), 50); // limit deducted

    const closeTx = await payer.closeTab(tab.id);
    assert.equal(closeTx.status, "confirmed");
    // No charges - full refund
    assert.equal(await payer.balance(), 100);
    assert.equal(await payee.balance(), 0);
  });

  it("tab throws when tab not found", async () => {
    const payer = mock.createWallet(100);
    await assert.rejects(() => payer.closeTab("nonexistent"), /TAB_NOT_FOUND|Tab not found/);
  });

  it("stream: open, close after time advance - correct accounting", async () => {
    const payer = mock.createWallet(1000);
    const payee = mock.createWallet(0);

    const stream = await payer.openStream({ to: payee.address, rate: 1, maxDuration: 100 });
    assert.equal(stream.status, "active");
    assert.equal(await payer.balance(), 900); // reserve of 100 deducted

    mock.advanceTime(30);
    await payer.closeStream(stream.id);

    const payerBal = await payer.balance();
    const payeeBal = await payee.balance();
    assert(payerBal > 960 && payerBal <= 970, `payer balance ${payerBal} should be ~970`);
    assert(payeeBal >= 30 && payeeBal <= 40, `payee balance ${payeeBal} should be ~30`);
  });

  it("bounty: post, award", async () => {
    const poster = mock.createWallet(100);
    const winner = mock.createWallet(0);
    const deadline = Math.floor(Date.now() / 1000) + 86400;

    const bounty = await poster.postBounty({ amount: 50, task: "Write tests", deadline });
    assert.equal(bounty.status, "open");
    assert.equal(await poster.balance(), 50);

    await winner.submitBounty(bounty.id, "0xevidence");
    await poster.awardBounty(bounty.id, 1);
    const b = mock.getBounty(bounty.id);
    assert.equal(b.status, "awarded");
    assert.equal(b.winner, winner.address);
    assert.equal(await winner.balance(), 50);
  });

  it("deposit: locks funds", async () => {
    const payer = mock.createWallet(100);
    const payee = mock.createWallet(0);

    const deposit = await payer.placeDeposit({ to: payee.address, amount: 30, expires: 3600 });
    assert.equal(deposit.status, "locked");
    assert.equal(await payer.balance(), 70);
  });

  it("status includes balance and tier", async () => {
    const wallet = mock.createWallet(555);
    const status = await wallet.status();
    assert.equal(parseFloat(status.balance), 555);
    assert(status.tier.length > 0);
  });

  it("reputation returns valid profile", async () => {
    const wallet = mock.createWallet();
    const rep = await wallet.getReputation(wallet.address);
    assert.equal(rep.address, wallet.address);
    assert(rep.score >= 0);
  });

  it("setBehavior forces next operation to fail", async () => {
    const payer = mock.createWallet(100);
    const payee = mock.createWallet(0);
    mock.setBehavior(payer.address, "INSUFFICIENT_BALANCE");
    await assert.rejects(() => payer.payDirect(payee.address, 1));
    // Second call should succeed (error was consumed)
    const tx = await payer.payDirect(payee.address, 1);
    assert.equal(tx.status, "confirmed");
  });

  it("advanceTime affects stream accounting", async () => {
    const payer = mock.createWallet(1000);
    const payee = mock.createWallet(0);

    const stream = await payer.openStream({ to: payee.address, rate: 10, maxDuration: 60 });
    mock.advanceTime(60); // full duration
    await payer.closeStream(stream.id);

    assert.equal(await payee.balance(), 600); // 10/s * 60s
    assert.equal(await payer.balance(), 400); // 1000 - 600
  });

  it("listBounties filters by status", async () => {
    const poster = mock.createWallet(1000);
    const w2 = mock.createWallet(0);
    const deadline = Math.floor(Date.now() / 1000) + 86400;

    const b1 = await poster.postBounty({ amount: 10, task: "Task A", deadline });
    await poster.postBounty({ amount: 10, task: "Task B", deadline });
    await w2.submitBounty(b1.id, "0xevidence");
    await poster.awardBounty(b1.id, 1);

    const open = mock.listBounties("open");
    const awarded = mock.listBounties("awarded");
    assert.equal(open.length, 1);
    assert.equal(awarded.length, 1);
  });

  it("requestTestnetFunds adds 100 USDC", async () => {
    const wallet = mock.createWallet(0);
    await wallet.requestTestnetFunds();
    assert.equal(await wallet.balance(), 100);
  });

  it("registerWebhook returns webhook object", async () => {
    const wallet = mock.createWallet();
    const wh = await wallet.registerWebhook("https://example.com/hook", ["payment.completed"]);
    assert.equal(wh.url, "https://example.com/hook");
    assert(wh.id.length > 0);
  });

  it("wallet toJSON does not expose private key", () => {
    const wallet = mock.createWallet();
    const json = JSON.stringify(wallet);
    // Should not contain anything resembling a private key
    assert(!json.includes("0x") || json.includes(wallet.address));
    assert(json.includes(wallet.address));
  });
});
